File: 算法/代码随想录算法PDF/00调试源代码/101.对称二叉树.py
#
# @lc app=leetcode.cn id=101 lang=python3
#
# [101] 对称二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        st = []
        st.append(root.left)
        st.append(root.right)
        while st:
            left = st.pop()
            right = st.pop()
            if not left and not right:
                continue
            if not left or not right or left.val != right.val:
                return False
            st.append(left.left)
            st.append(right.right)
            st.append(left.right)
            st.append(right.left)
        return True

# @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/101.对称二叉树.cpp
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (!root) return true;
        stack<TreeNode*> st;
        st.push(root->left);
        st.push(root->right);
        while (!st.empty()) {
            TreeNode* left = st.top();
            st.pop();
            TreeNode* right = st.top();
            st.pop();
            if (!left && !right) continue;  // 继续迭代
            if (!left || !right || left->val != right->val) {  // 不对称的情况
                return false;
            }
            st.push(left->left);  // 两两压栈
            st.push(right->right);
            st.push(left->right);
            st.push(right->left);
        }
        return true; // 正常遍历完没返回false，说明对称
    }
};
// @lc code=end

